### PR TITLE
[CRX] Migrate Chrome extension to Manifest Version 3

### LINKED ISSUE #15161: Plans for MV3?
@Rob--W 
Hi, I asked this question in the matrix channel and got told to put an issue here and ping you.

I just read your post about the Firefox add-on ecosystem here https://blog.mozilla.org/addons/2022/05/18/manifest-v3-in-firefox-recap-next-steps/

I understand that firefox add-ons will not be forced to migrate to service workers, but are there plans for migrating pdf.js to Chrome's MV3 requirements which do require this? Since the chrome web store has already stopped accepting mv2 extensions, it won't be possible to use pdf.js in any future chrome extensions without this effort, and it seems like by June 2023 pdf.js will simply stop working on chrome.

I wasn't able to find info about any of these plans online.

### PR DESCRIPTION
**Pull request by Rob--W**
This PR updates the implementation of the Chrome extension to Manifest Version 3. If you're interested in the individual units of change, review the individual commits and their commit messages.

I manually tested this in Chrome 127, 128 and 130 (and also as old as Chrome 96). Scenarios tested include: http(s) PDFs, link click (verifying that Referer is set), file:-navigations.

Fixes #15161

### CODE REVIEW

File: extensions/chromium/pdfHandler.js [hunk 1]
```diff
@@ -18,183 +18,243 @@ limitations under the License.
...
+    {
+      // Do not redirect for URLs containing pdfjs.action=download.
+      condition: {
+        urlFilter: "pdfjs.action=download",
+        resourceTypes: ["main_frame", "sub_frame"],
+      },
+      action: ACTION_IGNORE_OTHER_RULES,
+    },
+    {
+      // Do not redirect for URLs containing pdfjs.action=download.
```
Comment (timvandermeij) [inline extensions/chromium/pdfHandler.js:84]: This comment doesn't seem to match the condition below; is it perhaps accidentally copy/pasted from the one above?

File: extensions/chromium/pdfHandler.js [hunk 2]
```diff
@@ -18,183 +18,243 @@ limitations under the License.
...
+              "application/octet-stream;*",
+            ],
+          },
+        ],
+      },
+      action: ACTION_REDIRECT_TO_VIEWER,
+    },
+  ];
+  for (const [i, rule] of addRules.entries()) {
+    rule.id = 15161 + i;
```
Comment (timvandermeij) [inline extensions/chromium/pdfHandler.js:195]: Where does 15161 come from? Can we make it a constant to indicate its meaning?

File: extensions/chromium/preserve-referer.js [hunk 3]
```diff
@@ -60,9 +66,32 @@ function forgetReferrerEventually(tabId) {
...
+    delete g_postRequests[tabId];
   }, REFERRER_IN_MEMORY_TIME);
 }
 
+// Keeps track of whether a document in tabId + frameId is loaded through a
+// POST form submission. Although this logic has nothing to do with referrer
+// tracking, it is still here to enable re-use of the webRequest listener above.
+function setCanRequestBody(tabId, frameId, isPOST) {
+  if (isPOST) {
+    if (!g_postRequests[tabId]) {
```
Comment (timvandermeij) [inline extensions/chromium/preserve-referer.js:78]: For such cases (there are other occurrences in a previous commit for this file I believe) we could use the `||=` operator to shorten the code; see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR_assignment

**Review by timvandermeij — COMMENTED** [review]
Looks good to me, with the comments below addressed. I'll have another look at this once that's done before setting the final approval because it's quite a bit of code changes, so I'd like to make sure I haven't missed anything in this round.

File: extensions/chromium/pdfHandler.js (hunk 2 above)
Comment (Rob--W) [reply extensions/chromium/pdfHandler.js:195]: This is a "random" number, but fixed for predictability to make it easier to identify specific rules id ever needed. The value is the issue number linked this PR.

Every DNR ID needs to be unique, and be a positive number. So I would have to add at least 1 to the index. The exact value does not matter.

File: extensions/chromium/preserve-referer.js [hunk 4]
```diff
@@ -99,49 +85,50 @@ chrome.runtime.onConnect.addListener(function onReceivePort(port) {
...
-    }
-    chrome.webRequest.onBeforeSendHeaders.removeListener(onBeforeSendHeaders);
+    unsetStickyReferrer();
   });
-
-  function onBeforeSendHeaders(details) {
-    if (details.frameId !== frameId) {
-      return undefined;
+  function setStickyReferrer(url, callback) {
+    if (!referer) {
```
[the remark this answers is not in the document]
Comment (Rob--W) [reply extensions/chromium/preserve-referer.js:129]: The `referer`, `tabId`, `frameId` and `dnrRequestId` variables are all part of the closure. `let` vs `var` would not have made a difference here. I could move the `setStickyReferrer` function + `unsetStickyReferrer` functions to the top level to shrink the size of the `onConnect` handler here if you'd like.

File: extensions/chromium/pdfHandler.js (hunk 2 above)
Comment (Rob--W) [reply extensions/chromium/pdfHandler.js:195]: I'm changing to `+1` + a comment. The implementation clears any pre-existing dynamic rules if any, so the exact IDs do not matter, as long as they are unique and at least 1.

File: extensions/chromium/preserve-referer.js (hunk 3 above)
Comment (Rob--W) [reply extensions/chromium/preserve-referer.js:78]: While `||=` is the equivalent operator here, in this specific case the if-check was to check whether it was undefined vs `Set`. Thus I'm adapting your suggestion by using `??=` instead.

**Comment by Rob--W** [comment]
I've updated the code + rebased the PR. To more easily tell the real difference, here is the diff:

```diff
diff --git a/extensions/chromium/pdfHandler.js b/extensions/chromium/pdfHandler.js
index 0fd71e9f1..024be2760 100644
--- a/extensions/chromium/pdfHandler.js
+++ b/extensions/chromium/pdfHandler.js
@@ -81,8 +81,7 @@ async function registerPdfRedirectRule() {
       action: ACTION_IGNORE_OTHER_RULES,
     },
     {
-      // Redirect local PDF files if isAllowedFileSchemeAccess is true. No-op
-      // otherwise and then handled by webNavigation.onBeforeNavigate below.
+      // Do not redirect for URLs containing pdfjs.action=download.
       condition: {
         regexFilter: "^file://.*\\.pdf$",
         resourceTypes: ["main_frame", "sub_frame"],
@@ -193,8 +192,7 @@ async function registerPdfRedirectRule() {
     },
   ];
   for (const [i, rule] of addRules.entries()) {
-    // id must be unique and at least 1, but i starts at 0. So add +1.
-    rule.id = i + 1;
+    rule.id = 15161 + i;
     rule.priority = addRules.length - i;
   }
   try {
diff --git a/extensions/chromium/preserve-referer.js b/extensions/chromium/preserve-referer.js
index 92c628da3..d4da5b8f2 100644
--- a/extensions/chromium/preserve-referer.js
+++ b/extensions/chromium/preserve-referer.js
@@ -34,7 +34,7 @@ limitations under the License.
 
 // g_referrers[tabId][frameId] = referrer of PDF frame.
 var g_referrers = {};
-var g_referrerTimers = {};
+var g_referrerTimer = {};
 // The background script will eventually suspend after 30 seconds of inactivity.
 // This can be delayed when extension events are firing. To prevent the data
 // from being kept in memory for too long, cap the data duration to 5 minutes.
@@ -46,25 +46,26 @@ var g_postRequests = {};
 var rIsReferer = /^referer$/i;
 chrome.webRequest.onSendHeaders.addListener(
   function saveReferer(details) {
-    const { tabId, frameId, requestHeaders, method } = details;
-    g_referrers[tabId] ??= {};
-    g_referrers[tabId][frameId] = requestHeaders.find(h =>
-      rIsReferer.test(h.name)
+    if (!g_referrers[details.tabId]) {
+      g_referrers[details.tabId] = {};
+    }
+    g_referrers[details.tabId][details.frameId] = details.requestHeaders.find(
+      h => rIsReferer.test(h.name)
     )?.value;
-    setCanRequestBody(tabId, frameId, method !== "GET");
-    forgetReferrerEventually(tabId);
+    setCanRequestBody(details.tabId, details.frameId, details.method !== "GET");
+    forgetReferrerEventually(details.tabId);
   },
   { urls: ["*://*/*"], types: ["main_frame", "sub_frame"] },
   ["requestHeaders", "extraHeaders"]
 );
 
 function forgetReferrerEventually(tabId) {
-  if (g_referrerTimers[tabId]) {
-    clearTimeout(g_referrerTimers[tabId]);
+  if (g_referrerTimer[tabId]) {
+    clearTimeout(g_referrerTimer[tabId]);
   }
-  g_referrerTimers[tabId] = setTimeout(() => {
+  g_referrerTimer[tabId] = setTimeout(() => {
     delete g_referrers[tabId];
-    delete g_referrerTimers[tabId];
+    delete g_referrerTimer[tabId];
     delete g_postRequests[tabId];
   }, REFERRER_IN_MEMORY_TIME);
 }
@@ -74,7 +75,9 @@ function forgetReferrerEventually(tabId) {
 // tracking, it is still here to enable re-use of the webRequest listener above.
 function setCanRequestBody(tabId, frameId, isPOST) {
   if (isPOST) {
-    g_postRequests[tabId] ??= new Set();
+    if (!g_postRequests[tabId]) {
+      g_postRequests[tabId] = new Set();
+    }
     g_postRequests[tabId].add(frameId);
   } else {
     g_postRequests[tabId]?.delete(frameId);
```

File: extensions/chromium/preserve-referer.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (timvandermeij) [reply extensions/chromium/preserve-referer.js:126]: > I could move the setStickyReferrer function + unsetStickyReferrer functions to the top level to shrink the size of the onConnect handler here if you'd like.

Yes, moving the function definitions outside of the handler seems like a good idea here. Let's do that, and that'd be enough to address this comment.

**Comment by Rob--W** [comment]
Addressed https://github.com/mozilla/pdf.js/pull/18681#discussion_r1749228830

<details><summary>Output of: <code>git diff 0d9d9debd 4bf7be642</code></summary>

```diff
diff --git a/extensions/chromium/preserve-referer.js b/extensions/chromium/preserve-referer.js
index 92c628da3..b352dee36 100644
--- a/extensions/chromium/preserve-referer.js
+++ b/extensions/chromium/preserve-referer.js
@@ -112,7 +112,7 @@ chrome.runtime.onConnect.addListener(function onReceivePort(port) {
       referer = data.referer;
     }
     dnrRequestId = data.dnrRequestId;
-    setStickyReferrer(data.requestUrl, () => {
+    setStickyReferrer(dnrRequestId, tabId, data.requestUrl, referer, () => {
       // Acknowledge the message, and include the latest referer for this frame.
       port.postMessage(referer);
     });
@@ -120,41 +120,40 @@ chrome.runtime.onConnect.addListener(function onReceivePort(port) {
 
   // The port is only disconnected when the other end reloads.
   port.onDisconnect.addListener(function () {
-    unsetStickyReferrer();
+    unsetStickyReferrer(dnrRequestId);
   });
-  function setStickyReferrer(url, callback) {
-    if (!referer) {
-      unsetStickyReferrer();
-      callback();
-      return;
-    }
-    const rule = {
-      id: dnrRequestId,
-      condition: {
-        urlFilter: `|${url}|`,
-        // The viewer and background are presumed to have the same origin:
-        initiatorDomains: [location.hostname], // = chrome.runtime.id.
-        resourceTypes: ["xmlhttprequest"],
-        tabIds: [tabId],
-      },
-      action: {
-        type: "modifyHeaders",
-        requestHeaders: [
-          { operation: "set", header: "referer", value: referer },
-        ],
-      },
-    };
-    chrome.declarativeNetRequest.updateSessionRules(
-      { removeRuleIds: [dnrRequestId], addRules: [rule] },
-      callback
-    );
+});
+
+function setStickyReferrer(dnrRequestId, tabId, url, referer, callback) {
+  if (!referer) {
+    unsetStickyReferrer(dnrRequestId);
+    callback();
+    return;
   }
+  const rule = {
+    id: dnrRequestId,
+    condition: {
+      urlFilter: `|${url}|`,
+      // The viewer and background are presumed to have the same origin:
+      initiatorDomains: [location.hostname], // = chrome.runtime.id.
+      resourceTypes: ["xmlhttprequest"],
+      tabIds: [tabId],
+    },
+    action: {
+      type: "modifyHeaders",
+      requestHeaders: [{ operation: "set", header: "referer", value: referer }],
+    },
+  };
+  chrome.declarativeNetRequest.updateSessionRules(
+    { removeRuleIds: [dnrRequestId], addRules: [rule] },
+    callback
+  );
+}
 
-  function unsetStickyReferrer() {
-    if (dnrRequestId) {
-      chrome.declarativeNetRequest.updateSessionRules({
-        removeRuleIds: [dnrRequestId],
-      });
-    }
+function unsetStickyReferrer(dnrRequestId) {
+  if (dnrRequestId) {
+    chrome.declarativeNetRequest.updateSessionRules({
+      removeRuleIds: [dnrRequestId],
+    });
   }
-});
+}

```
</details>

Or perhaps a bit easier to review:

<details><summary>Output of: <code>git diff 0d9d9debd 4bf7be642 --ignore-space-change</code></summary>

```diff
diff --git a/extensions/chromium/preserve-referer.js b/extensions/chromium/preserve-referer.js
index 92c628da3..b352dee36 100644
--- a/extensions/chromium/preserve-referer.js
+++ b/extensions/chromium/preserve-referer.js
@@ -112,7 +112,7 @@ chrome.runtime.onConnect.addListener(function onReceivePort(port) {
       referer = data.referer;
     }
     dnrRequestId = data.dnrRequestId;
-    setStickyReferrer(data.requestUrl, () => {
+    setStickyReferrer(dnrRequestId, tabId, data.requestUrl, referer, () => {
       // Acknowledge the message, and include the latest referer for this frame.
       port.postMessage(referer);
     });
@@ -120,11 +120,13 @@ chrome.runtime.onConnect.addListener(function onReceivePort(port) {
 
   // The port is only disconnected when the other end reloads.
   port.onDisconnect.addListener(function () {
-    unsetStickyReferrer();
+    unsetStickyReferrer(dnrRequestId);
   });
-  function setStickyReferrer(url, callback) {
+});
+
+function setStickyReferrer(dnrRequestId, tabId, url, referer, callback) {
   if (!referer) {
-      unsetStickyReferrer();
+    unsetStickyReferrer(dnrRequestId);
     callback();
     return;
   }
@@ -139,22 +141,19 @@ chrome.runtime.onConnect.addListener(function onReceivePort(port) {
     },
     action: {
       type: "modifyHeaders",
-        requestHeaders: [
-          { operation: "set", header: "referer", value: referer },
-        ],
+      requestHeaders: [{ operation: "set", header: "referer", value: referer }],
     },
   };
   chrome.declarativeNetRequest.updateSessionRules(
     { removeRuleIds: [dnrRequestId], addRules: [rule] },
     callback
   );
-  }
+}
 
-  function unsetStickyReferrer() {
+function unsetStickyReferrer(dnrRequestId) {
   if (dnrRequestId) {
     chrome.declarativeNetRequest.updateSessionRules({
       removeRuleIds: [dnrRequestId],
     });
   }
-  }
-});
+}
```
</details>

File: extensions/chromium/preserve-referer.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Rob--W) [reply extensions/chromium/preserve-referer.js]: Done: https://github.com/mozilla/pdf.js/pull/18681#issuecomment-2336711233

**Review by timvandermeij — APPROVED** [review]

**Comment by timvandermeij** [comment]
Nice work; thank you!

**Pull request merged**